# Part 1: Forward Mode Automatic Differentiation

Forward mode AD can simply be implemented by defining a class to represent [dual numbers](https://en.wikipedia.org/wiki/Dual_number) which hold the value and its derivative. The following skeleton defines a dual number and implements multiplication. 

__Tasks:__

- Addition (`__add__`) is incomplete - can you finish it? 
- Can you also implement division (`__truediv__`), subtraction (`__sub__`) and power (`__pow__`)?

In [0]:
import math

class DualNumber:
    def __init__(self, value, dvalue):
        self.value = value
        self.dvalue = dvalue

    def __str__(self):
        return str(self.value) + " + " + str(self.dvalue) + "ε"

    def __mul__(self, other):
        return DualNumber(self.value * other.value,
            self.dvalue * other.value + other.dvalue * self.value)
    
    def __add__(self, other):
        #TODO: finish me
        # YOUR CODE HERE
        return DualNumber(self.value+other.value,
                          self.dvalue+other.dvalue)
    
    # TODO: add missing methods
    # YOUR CODE HERE
    def __truediv__(self,other):
        return DualNumber(self.value/other.value,
                          (self.dvalue*other.value-self.value*other.dvalue)/other.value**2)
    
    def __sub__(self,other):
        return DualNumber(self.value-other.value,
                          self.dvalue-other.dvalue)
      
    def __pow__(self,other):
        return DualNumber(self.value**2,
                          2*self.value*self.dvalue)
    

In [0]:
# Tests

a=DualNumber(1,0) + DualNumber(1,0) / DualNumber(1,0) - DualNumber(1,0)**DualNumber(1,0)
print(a)

1.0 + 0.0ε


## Implementing math functions

We also need to implement some core math functions. Here's the sine function for a dual number:

In [0]:
def sin(x):
    return DualNumber(math.sin(x.value), math.cos(x.value)*x.dvalue)

__Task:__ can you implement the _cosine_ (`cos`), _tangent_ (`tan`), and _exponential_ (`exp`) functions in the code block below?

In [0]:
# TODO: implement additional math functions on dual numbers

def cos(x):
    # YOUR CODE HERE
#     raise NotImplementedError()
    return DualNumber(math.cos(x.value),
                      -x.dvalue*math.sin(x.value))

def tan(x):
    # YOUR CODE HERE
#     raise NotImplementedError()
    return DualNumber(math.tan(x.value),
                     x.dvalue/math.cos(x.value)**2)

def exp(x):
    # YOUR CODE HERE
#     raise NotImplementedError()
    return DualNumber(math.exp(x.value),
                     x.dvalue*math.exp(x.value))

In [0]:
# Tests
assert cos(DualNumber(0,0)).value == 1
assert tan(DualNumber(0,0)).value == 0
assert exp(DualNumber(0,0)).value == 1


## Time to try it out

We're now in a position to try our implementation.

__Task:__ 

- Try running the following code to compute the value of the function $z=x\cdot y+sin(x)$ given $x=0.5$ and $y=4.2$, together with the derivative $\partial z/\partial x$ at that point. 

In [0]:
# YOUR CODE HERE
# raise NotImplementedError()
x=DualNumber(0.5,1)
y=DualNumber(4.2,0)
z=x*y+sin(x)
print("z =",z.value)
print("∂z/∂x = ",z.dvalue)

z = 2.579425538604203
∂z/∂x =  5.077582561890373


__Task__: Differentiate the above function with respect to $x$ and write the symbolic derivatives in the following box. Verify the result computed above is correct by plugging-in the values into your symbolic gradient expression.

In [0]:
4.2+math.cos(0.5)

5.077582561890373

**Answer**:

$\partial z/\partial x = y+cos(x)$ at point $x=0.5$ any $y=4.2$ is $4.2+cos(0.5)=5.07758$

__Task:__ Now use the code block below to compute the derivative $\partial z/\partial y$ of the above expression (at the same point $x=0.5, y=4.2$ as above) and store the derivative in the variable `dzdy` (just the derivative, not the Dual Number). Verify by hand that the result is correct.

In [0]:
# YOUR CODE HERE
# raise NotImplementedError()
x=DualNumber(0.5,0)
y=DualNumber(4.2,1)
z=x*y+sin(x)
dzdy=z.dvalue
print('dz/dy:', dzdy)

dz/dy: 0.5


**Answer**:

$\partial z/\partial y = x$ at point $x=0.5$ any $y=4.2$ is $0.5$

In [0]:
#Tests
assert dzdy == 0.5
assert type(dzdy) == float


__Task:__ Finally, use the code block below to experiment and test the other math functions and methods you created.

In [0]:
x = DualNumber(0.5,1)
z = x ** 2
print('z:', z)

print('∂z/∂x:',z.dvalue)

z: 0.25 + 1.0ε
∂z/∂x: 1.0


In [0]:
x = DualNumber(2,1)
y = DualNumber(4,0)
z = x / y
print('z:', z)
print('∂z/∂x:',z.dvalue)

x = DualNumber(2,0)
y = DualNumber(4,1)
z = x / y
print('z:', z)
print('∂z/∂y:',z.dvalue)

z: 0.5 + 0.25ε
∂z/∂x: 0.25
z: 0.5 + -0.125ε
∂z/∂y: -0.125


compute the value of the function $z=x/y+x^2+y^2+cos(x)+tan(x)-exp(x)$ given $x=1$ and $y=2$, together with the derivative $\partial z/\partial x$ and  $\partial z/\partial y$ at that point. 

In [0]:
# YOUR CODE HERE
# raise NotImplementedError()
x=DualNumber(1,1)
y=DualNumber(2,0)
z=x/y+x**x+y**y+cos(x)+tan(x)-exp(x)
print("z =",z.value)
print("∂z/∂x = ",z.dvalue)

z = 4.879428202063998
∂z/∂x =  2.3657660075478177


In [0]:
x=DualNumber(1,0)
y=DualNumber(2,1)
z=x/y+x**x+y**y+cos(x)+tan(x)-exp(x)
print("∂z/∂y = ",z.dvalue)

∂z/∂y =  3.75
